### PR TITLE
Check for duplicate YAML keys when linting

### PR DESCRIPTION
**Pull request by FrNecas**
Resolves: #511

### CODE REVIEW

File: tests/test/lint/test.sh [hunk 1]
```diff
@@ -34,6 +34,8 @@ rlJournalStart
         rlAssertGrep 'fail directory path must exist' output
         rlRun "tmt test lint relevancy | tee output" 1
         rlAssertGrep 'fail relevancy has been obsoleted' output
+        rlRun "tmt test lint duplicate | tee output" 1
+        rlAssertgrep 'fail fmf metadata must not contain duplicate keys' output
```
Comment (lukaszachy) [inline tests/test/lint/test.sh:38]:
```suggestion
        rlAssertGrep 'fail fmf metadata must not contain duplicate keys' output
```

File: tmt/base.py [hunk 2]
```diff
@@ -364,7 +365,11 @@ def lint(self):
 
         # Check for possible test case relevancy rules
         filename = self.node.sources[-1]
```
Comment (lukaszachy) [inline tmt/base.py:367]: Hm, this is a problem - Duplicate can be created by parent node. E.g.
```
==> main.fmf <==
test: y
test: y

==> x.fmf <==
a: b
```
And as such 
```
$ tmt tests lint

/x
pass test script must be defined
pass directory path must be absolute
pass directory path must exist
warn summary is very useful for quick inspection
```
However when I remove `x.fmf` file:
```
/
pass test script must be defined
pass directory path must be absolute
pass directory path must exist
warn summary is very useful for quick inspection
fail fmf metadata must not contain duplicate keys
```

**Review by lukaszachy — CHANGES_REQUESTED** [review]

**Comment by psss** [comment]
What if we cover the check for duplication already in `fmf`?

**Comment by FrNecas** [comment]
If we want to check for duplicates while considering inheritance, then this really belongs to `fmf` (perhaps we should introduce a new exception there for such an error) and `tmt` should only catch the exception. I will leave this PR open for now, hopefully modify fmf and then add some test coverage to tmt for catching the exception.

I am not completely sure why `fmf` isn't used to read the data in `lint`. Seems like it was introduced by this commit https://github.com/psss/tmt/commit/d94f6019bfe2f5b11e42eb9f85ce94815653d6f6 . @psss could you perhaps provide some more insight into this?

**Comment by psss** [comment]
> perhaps we should introduce a new exception there for such an error and `tmt` should only catch the exception

Yes, that makes a perfect sense to me.

> I am not completely sure why fmf isn't used to read the data in lint

The part implementing the `tmt test lint --fix` functionality needed to write data to disk and there was no support to store data from `fmf`. A very simple support for writing modified metadata to disk has been merged in psss/fmf/pull/102 just yesterday.

**Comment by FrNecas** [comment]
Thanks for the explanation with `--fix`. I've tried checking for the duplicate keys in fmf and found a potential issue. Lets take as an example `tmt test lint`, if we raise an Exception when constructing the fmf tree, the data won't be complete and this line is going to fail: https://github.com/psss/tmt/blob/master/tmt/cli.py#L366 and hence we can only really output an error that there are duplicate keys and no other linting will happen. So I see 2 options:

a) Check for duplicates in fmf (and support inheritance) and be OK with (not only) test linting immediately failing if there is a duplicate key (I am not a big fan of this one)
b) Only check the "local" YAML for duplicates (like in the current form of this PR), the linting goes through fine but it doesn't take into account the inheritance hierarchy.

One more possible alternative that I think could work (but I am not an expert on fmf so not sure) - ignore duplicates in fmf when first parsing the tree (so the aforementioned loop would go through normally) but somehow allow reading the tree data while also checking for duplicates (a method which we could call in `Test.lint()` to check that there are no duplicate keys). What do you think?

**Comment by psss** [comment]
My recommended approach would be that `fmf` screams as soon as it runs into a `yaml` with duplicate keys. If the raised error contains the key culprit name and a full path to the file, that is the only thing a user needs to get it fixed. I don't think it's worth to invest much time to present a nice linting experience for these errors. Plus, if we cover this in `fmf` we don't have to lint various other places where a duplicate key can appear.

**Comment by FrNecas** [comment]
OK, sounds good. I will open a PR to fmf with the necessary changes so I am closing this one, no changes will be necessary on tmt side because an error when parsing YAML is going to result in a `FileError` which is already taken care of here.

**Pull request closed without merging**